### PR TITLE
Upgrade PHPUnit to 9.3

### PR DESCRIPTION
**Pull request by lcobucci**
|    Q          |   A
|-------------- | ------
| New Feature   | yes

### Description

This applies all necessary changes to make sure PHPUnit is upgraded to v9.3.
Tests were drastically refactored to modernise them and remove some bad practices.

Related to #50

### CODE REVIEW

File: .phpunit.result.cache [hunk 1]
```diff
@@ -0,0 +1 @@
...thAllHttpMethods with data set "laminas-OPTIONS"";d:0.001;s:99:"MezzioTest\Router\IntegrationTest::testAllowedMethodsWhenOnlyPutMethodSet with data set "aura-head"";d:0.001;s:102:"MezzioTest\Router\IntegrationTest::testAllowedMethodsWhenOnlyPutMethodSet with data set "aura-options"";d:0.001;s:105:"MezzioTest\Router\IntegrationTest::testAllowedMethodsWhenOnlyPutMethodSet with data set "fast-route-head"";d:0.001;s:108:"MezzioTest\Router\IntegrationTest::testAllowedMethodsWhenOnlyPutMethodSet with data set "fast-route-options"";d:0.001;s:102:"MezzioTest\Router\IntegrationTest::testAllowedMethodsWhenOnlyPutMethodSet with data set "laminas-head"";d:0.002;s:105:"MezzioTest\Router\IntegrationTest::testAllowedMethodsWhenOnlyPutMethodSet with data set "laminas-options"";d:0.002;s:119:"MezzioTest\Router\IntegrationTest::testWithOnlyRootPathRouteDefinedRoutingToSubPathsShouldDelegate with data set "aura"";d:0.001;s:125:"MezzioTest\Router\IntegrationTest::testWithOnlyRootPathRouteDefinedRoutingToSubPathsShouldDelegate with data set "fast-route"";d:0.001;s:122:"MezzioTest\Router\IntegrationTest::testWithOnlyRootPathRouteDefinedRoutingToSubPathsShouldDelegate with data set "laminas"";d:0.002;}}}
```
Comment (jensdenies) [inline .phpunit.result.cache]: I think this one accidentally slipped through. 😄

Comment (lcobucci) [reply .phpunit.result.cache]: That's why I shouldn't be coding after midnight :rofl:

**Comment by jensdenies** [comment]
@lcobucci What workflow do you use to upgrade dependencies, do you prefer the lowest available version in composer for the supported PHP version or the highest?

File: composer.json [hunk 2]
```diff
@@ -51,9 +51,11 @@
...
         "mezzio/mezzio-laminasrouter": "^3.0",
         "mockery/mockery": "^1.0",
-        "phpstan/phpstan": "^0.9.2",
-        "phpstan/phpstan-strict-rules": "^0.9",
-        "phpunit/phpunit": "^7.0.1"
+        "phpstan/extension-installer": "^1.1",
+        "phpstan/phpstan": "^0.12",
+        "phpstan/phpstan-phpunit": "^0.12",
+        "phpstan/phpstan-strict-rules": "^0.12",
+        "phpunit/phpunit": "^9.3"
```
Comment (rieschl) [inline composer.json:58]: Hm, I thought, the TSC decided to use psalm, and not phpstan?

Comment (Ocramius) [reply composer.json:58]: Yeah, but the tooling was pre-installed/running, so it makes sense to upgrade them for now, and do the psalm migration separately.

**Comment by lcobucci** [comment]
@Ocramius @boesing would mind reviewing this one? Or should I simply get it merged?

**Review by Ocramius — APPROVED** [review]
LGTM 👍

**Pull request merged**